### PR TITLE
RtpStreamSend: Do not store too old RTP packets

### LINKED ISSUE #882: Crash at assertion 'StorageItemBuffer contains more than 65535 entries'
## Bug Report
It occurred once in three days on ten servers.
```
(ABORT) RTC::RtpStreamSend::Insert() | failed assertion `this->buffer.size() <= MaxSeq': StorageItemBuffer contains more than 65535 entries
```

### Your environment

- Operating system: Ubuntu 20.04.2
- Node version:
- npm version:
- gcc/clang version:
- mediasoup version:Rust version mediasoup: 0.10.0 mediasoup-sys: 0.4.2
- mediasoup-client version: 3.6.50

### Issue description
==== updated ====
At the time of the crash, there were  6 producers and 17 consumers.
It may have continued to produce for a long time because the number of producers did not increase or decrease during the low user time period. (I do not know for sure, as I do not have produce time metrics.)

### PR DESCRIPTION
**Pull request by jmillan**
Fixes #882

### CODE REVIEW

File: worker/src/RTC/RtpPacket.cpp [hunk 1]
```diff
@@ -117,10 +117,7 @@ namespace RTC
 		           payloadLength + size_t{ payloadPadding },
 		  "packet's computed size does not match received size");
 
-		auto* packet =
-		  new RtpPacket(header, headerExtension, payload, payloadLength, payloadPadding, len);
-
-		return packet;
+		return new RtpPacket(header, headerExtension, payload, payloadLength, payloadPadding, len);
```
Comment (jmillan) [inline worker/src/RTC/RtpPacket.cpp:120]: Make fuzzer happy. It was complaining about a leak.

File: worker/src/RTC/RtpStreamSend.cpp [hunk 2]
```diff
@@ -452,6 +457,23 @@ namespace RTC
 			return;
 		}
 
+		// Check if RTP packet is too old to be stored.
```
Comment (jmillan) [inline worker/src/RTC/RtpStreamSend.cpp:460]: This is the fix.

Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:460]: On one side we were clearing old packets and on the other we were adding old ones that would later lead to inconsistencies.

File: worker/src/RTC/RtpStreamSend.cpp [hunk 3]
```diff
@@ -502,16 +523,11 @@ namespace RTC
 		// items that contain packets older than `MaxRetransmissionDelay`.
 		for (size_t i{ 0 }; i < bufferSize && this->storageItemBuffer.GetBufferSize() != 0; ++i)
 		{
-			auto* firstStorageItem = this->storageItemBuffer.GetFirst();
-
-			MS_ASSERT(firstStorageItem, "first storage item is missing");
-			MS_ASSERT(firstStorageItem->packet, "storage item does not contain original packet");
```
Comment (jmillan) [inline worker/src/RTC/RtpStreamSend.cpp:508]: moved to `GetFirst()`

File: worker/src/RTC/RtpStreamSend.cpp [hunk 4]
```diff
@@ -502,16 +523,11 @@ namespace RTC
...
-
-			auto firstPacketTs{ firstStorageItem->timestamp };
-			uint32_t diffTs{ packetTs - firstPacketTs };
+			auto* storageItem = this->storageItemBuffer.GetFirst();
```
Comment (jmillan) [inline worker/src/RTC/RtpStreamSend.cpp:525]: Changes in this method are cosmetic.

**Comment by jmillan** [comment]
Fuzzer would detect the crash right away, after the change it runs for long minutes without issues.

File: worker/src/RTC/RtpStreamSend.cpp [hunk 5]
```diff
@@ -452,6 +457,23 @@ namespace RTC
...
+
+			// Processing RTP packet is older than first one.
+			if (RTC::SeqManager<uint32_t>::IsSeqLowerThan(packet->GetTimestamp(), storageItem->timestamp))
+			{
+				uint32_t diffTs{ storageItem->timestamp - packet->GetTimestamp() };
+
+				// RTP packet is older than the retransmission buffer size.
+				if (static_cast<uint32_t>(diffTs * 1000 / this->params.clockRate) >= this->retransmissionBufferSize)
+
```
Comment (ibc) [inline worker/src/RTC/RtpStreamSend.cpp]: Annoying blank line or missing brackets?

Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp]: The former, fixed.

**Review by nazar-pc — APPROVED** [review]
Makes sense to me, bug was tricky, glad that we have fuzzer

**Comment by jmillan** [comment]
I'm merging and releasing a new version

**Comment by nazar-pc** [comment]
Can you bump `mediasoup-sys` patch version so I can publish Rust release too?

**Comment by jmillan** [comment]
> Can you bump mediasoup-sys patch version so I can publish Rust release too?

It's only about the `version` field in Cargo.toml, right?

**Comment by nazar-pc** [comment]
Yes, just patch version in `worker/Cargo.toml`

**Review by ibc — COMMENTED** [review]
Please add in CHANGELOG

**Comment by jmillan** [comment]
You are very right, I missed that and I already released..

**Comment by jmillan** [comment]
Done

**Pull request merged**